### PR TITLE
upgrade python to 3.12

### PR DESCRIPTION
**Pull request by sastels**
# Summary | Résumé

Upgrade python.

## Related Issues | Cartes liées

* https://app.zenhub.com/workspaces/notify-planning-614b3ad91bc2030015ed22f5/issues/gh/cds-snc/notification-document-download-api/195
* https://app.zenhub.com/workspaces/notify-planning-core-6411dfb7c95fb80014e0cab0/issues/gh/cds-snc/notification-planning-core/355

# Test instructions | Instructions pour tester la modification

rebuild and run document-download-api locally together with `main` api.

# Release Instructions | Instructions pour le déploiement

None.

# Reviewer checklist | Liste de vérification du réviseur

- [ ] This PR does not break existing functionality.
- [ ] This PR does not violate GCNotify's privacy policies.
- [ ] This PR does not raise new security concerns. Refer to our GC Notify Risk Register document on our Google drive.
- [ ] This PR does not significantly alter performance.
- [ ] Additional required documentation resulting of these changes is covered (such as the README, setup instructions, a related ADR or the technical documentation).

> ⚠ If boxes cannot be checked off before merging the PR, they should be moved to the "Release Instructions" section with appropriate steps required to verify before release. For example, changes to celery code may require tests on staging to verify that performance has not been affected.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -28,7 +28,7 @@ gevent = "24.2.1"
 gunicorn = "22.0.0"
 
 
-notifications-utils = { git = "https://github.com/cds-snc/notifier-utils.git", tag = "52.2.7"}
```
Comment (sastels) [inline pyproject.toml:31]: Change this after the utils PR merges

File: pyproject.toml [hunk 2]
```diff
@@ -28,7 +28,7 @@ gevent = "24.2.1"
...
+notifications-utils = { git = "https://github.com/cds-snc/notifier-utils.git", branch = "python-3.12"}
```
Comment (sastels) [inline pyproject.toml:31]: Change this after the utils PR merges

File: Makefile [hunk 3]
```diff
@@ -6,6 +6,10 @@ GIT_COMMIT ?= $(shell git rev-parse HEAD)
 run:
 	FLASK_APP=application.py FLASK_ENV=development poetry run flask run -p 7000
 
+.PHONY: run-gunicorn
+run-gunicorn:
+	PORT=7000 poetry run gunicorn -c gunicorn_config.py application
+
```
Comment (sastels) [inline Makefile:12]: Just added to help test gunicorn locally

File: ci/Dockerfile [hunk 4]
```diff
@@ -1,4 +1,4 @@
-FROM python:3.10-alpine3.18@sha256:d5ee9613c89c9bd4c4112465d2136512ea8629bce6ff15fa27144f3cc16b5c6b
+FROM python:3.12-alpine3.20
```
Comment (ben851) [inline ci/Dockerfile:1]: Can we use the sha here as well?

**Review by ben851 — CHANGES_REQUESTED** [review]
Can we use the sha on the docker image to remain consistent/secure

File: ci/Dockerfile (hunk 4 above)
Comment (ben851) [reply ci/Dockerfile]: that I did not know!

**Review by ben851 — APPROVED** [review]
LGTM

**Pull request merged**